rspace/src/main/scala/coop/rchain/rspace/ReplayRSpace.scala

```diff
@@ -0,0 +1,305 @@
+
+        def handleMatch(mat: ProduceCandidate[C, P, A, K],
+                        replays: ReplayData,
+                        produceRef: Produce,
+                        comms: Multiset[COMM]): Option[(K, Seq[A])] =
+          mat match {
+            case ProduceCandidate(channels,
+                                  WaitingContinuation(_, continuation, persistK, consumeRef),
+                                  continuationIndex,
+                                  dataCandidates) =>
+              store.eventsCounter.registerProduceCommEvent()
+              val commRef = COMM(consumeRef, dataCandidates.map(_.datum.source))
+              assert(comms.contains(commRef), "COMM Event was not contained in the trace")
+              if (!persistK) {
+                store.removeWaitingContinuation(txn, channels, continuationIndex)
+              }
+              dataCandidates
+                .sortBy(_.datumIndex)(Ordering[Int].reverse)
+                .foreach {
+                  case DataCandidate(candidateChannel, Datum(_, persistData, prodRef), dataIndex) =>
+                    if (!persistData && dataIndex >= 0) {
+                      store.removeDatum(txn, Seq(candidateChannel), dataIndex)
+                    }
+                    store.removeJoin(txn, candidateChannel, channels)
+                    replays.removeBinding(prodRef, commRef)
+                }
+              logger.debug(s"produce: matching continuation found at <channels: $channels>")
+              replayData.put(replays.removeBinding(consumeRef, commRef))
+              Some((continuation, dataCandidates.map(_.datum.a)))
+          }
+
+        val groupedChannels: Seq[Seq[C]] = store.getJoin(txn, channel)
+
+        logger.debug(s"""|produce: searching for matching continuations
+                         |at <groupedChannels: $groupedChannels>""".stripMargin.replace('\n', ' '))
+
+        val produceRef = Produce.create(channel, data, persist)
+        val replays    = replayData.take()
+
+        replays.get(produceRef) match {
+          case None =>
+            runMatcher(None, produceRef, groupedChannels) match {
+              case None =>
+                storeDatum(replays, produceRef)
+              case Some(ProduceCandidate(channels, _, _, _)) =>
+                logger.debug(s"produce: matching continuation found at <channels: $channels>")
+                val msg = "untraced event resulted in a comm event"
+                logger.error(msg)
+                throw new ReplayException(msg)
+            }
+          case Some(comms) =>
+            val maybeProduceCandidate: Option[ProduceCandidate[C, P, A, K]] =
+              comms
+                .iterator()
+                .asScala
+                .map { comm: COMM =>
+                  runMatcher(Some(comm), produceRef, groupedChannels)
+                }
+                .collectFirst {
+                  case Some(x) => x
+                }
+
+            maybeProduceCandidate match {
```

Comment:
Can we get the rename "maybeCandidateConsume" or are you planning to do it all at once later?

Reply:
I'd prefer to address renaming later, as it will span across both RSpaces...

Reply:
Got it